File: src/dotnet/Harness.Engine/PlanRevisionStore.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Harness.Engine;

/// <summary>Persists proposed and accepted global plan revisions for audit and resume.</summary>
public static class PlanRevisionStore
{
    public const string ProposalPath = ".harness/replan.json";
    private const string Dir = ".harness/plans";
    private const string CurrentPath = ".harness/plan_revision.json";

    public static PlanRevision? ReadProposal()
    {
        try
        {
            if (!File.Exists(ProposalPath)) return null;
            return JsonSerializer.Deserialize(File.ReadAllText(ProposalPath), HarnessJsonContext.Default.PlanRevision);
        }
        catch (Exception ex)
        {
            HarnessLog.Error($"[PlanRevisionStore] failed to read proposal: {ex.Message}");
            return null;
        }
    }

    public static int RevisionCount()
    {
        try
        {
            if (!File.Exists(CurrentPath)) return 0;
            return JsonSerializer.Deserialize(File.ReadAllText(CurrentPath), HarnessJsonContext.Default.AppliedPlanRevision)?.Version ?? 0;
        }
        catch { return 0; }
    }

    public static void Record(
        PlanRevision revision, IReadOnlyList<Feature> features, PlanRevisionEvaluation evaluation)
    {
        var applied = new AppliedPlanRevision(
            RevisionCount() + 1,
            DateTimeOffset.UtcNow,
            revision.Reason,
            revision.Alternatives,
            revision.ObservationIds,
            Fingerprint(features),
            evaluation,
            [.. features]);
        Directory.CreateDirectory(Dir);
        var json = JsonSerializer.Serialize(applied, PrettyFeatureListJsonContext.Default.AppliedPlanRevision);
        AtomicIO.WriteAllTextAtomic(CurrentPath, json);
        AtomicIO.WriteAllTextAtomic(Path.Combine(Dir, $"plan-v{applied.Version}.json"), json);
    }

    public static bool HasPlanFingerprint(string fingerprint)
    {
        if (!Directory.Exists(Dir)) return false;
        foreach (var path in Directory.EnumerateFiles(Dir, "plan-v*.json"))
        {
            try
            {
                var applied = JsonSerializer.Deserialize(
                    File.ReadAllText(path), HarnessJsonContext.Default.AppliedPlanRevision);
                if (applied?.PlanFingerprint == fingerprint) return true;
            }
            catch { /* corrupt history never approves a proposal */ }
        }
        return false;
    }

    public static string Fingerprint(IReadOnlyList<Feature> features)
    {
        var canonical = features.OrderBy(f => f.Id).Select(f => f with { Passes = false }).ToList();
        var json = JsonSerializer.Serialize(new FeatureList(canonical), HarnessJsonContext.Default.FeatureList);
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(json))).ToLowerInvariant();
    }

    public static void Reset()
    {
        try
        {
            if (File.Exists(ProposalPath)) File.Delete(ProposalPath);
            if (File.Exists(CurrentPath)) File.Delete(CurrentPath);
            if (Directory.Exists(Dir)) Directory.Delete(Dir, recursive: true);
        }
        catch (Exception ex)
        {
            HarnessLog.Error($"[PlanRevisionStore] failed to reset: {ex.Message}");
        }
    }
}

public sealed record AppliedPlanRevision(
    int Version,
    DateTimeOffset AppliedAt,
    string Reason,
    string[] AlternativesConsidered,
    string[] BasedOnObservationIds,
    string PlanFingerprint,
    PlanRevisionEvaluation Evaluation,
    List<Feature> Features);

File: src/go/harnessengine/plan_revision_store.go
package harnessengine

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// Persists proposed and accepted global plan revisions for audit and resume.
const (
	// ReplanProposalPath is where the driver writes its proposed revision — read by
	// ReadPlanRevisionProposal, the same way plan.json is written for `plan`.
	ReplanProposalPath = ".harness/replan.json"

	planRevisionsDir        = ".harness/plans"
	currentPlanRevisionPath = ".harness/plan_revision.json"
)

// AppliedPlanRevision is the audit record of one accepted plan revision.
type AppliedPlanRevision struct {
	Version                int                    `json:"version"`
	AppliedAt              time.Time              `json:"appliedAt"`
	Reason                 string                 `json:"reason"`
	AlternativesConsidered []string               `json:"alternativesConsidered"`
	BasedOnObservationIds  []string               `json:"basedOnObservationIds"`
	PlanFingerprint        string                 `json:"planFingerprint"`
	Evaluation             PlanRevisionEvaluation `json:"evaluation"`
	Features               []Feature              `json:"features"`
}

// ReadPlanRevisionProposal reads the driver-written proposal at ReplanProposalPath. nil if
// absent/unreadable/unparseable.
func ReadPlanRevisionProposal() *PlanRevision {
	if !fileExists(ReplanProposalPath) {
		return nil
	}
	data, err := os.ReadFile(ReplanProposalPath)
	if err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to read proposal: %s", err))
		return nil
	}
	var revision PlanRevision
	if err := json.Unmarshal(data, &revision); err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to read proposal: %s", err))
		return nil
	}
	return &revision
}

// PlanRevisionCount returns the Version of the currently applied revision, 0 if none.
func PlanRevisionCount() int {
	if !fileExists(currentPlanRevisionPath) {
		return 0
	}
	data, err := os.ReadFile(currentPlanRevisionPath)
	if err != nil {
		return 0
	}
	var applied AppliedPlanRevision
	if err := json.Unmarshal(data, &applied); err != nil {
		return 0
	}
	return applied.Version
}

// RecordPlanRevision persists an accepted revision as both the current pointer
// (.harness/plan_revision.json) and a numbered history entry (.harness/plans/plan-v{N}.json).
func RecordPlanRevision(revision PlanRevision, features []Feature, evaluation PlanRevisionEvaluation) {
	persisted := make([]Feature, len(features))
	copy(persisted, features)

	applied := AppliedPlanRevision{
		Version:                PlanRevisionCount() + 1,
		AppliedAt:              time.Now().UTC(),
		Reason:                 revision.Reason,
		AlternativesConsidered: revision.Alternatives,
		BasedOnObservationIds:  revision.ObservationIds,
		PlanFingerprint:        PlanRevisionFingerprint(features),
		Evaluation:             evaluation,
		Features:               persisted,
	}

	if err := ensureDir(planRevisionsDir); err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to record: %s", err))
		return
	}
	data, err := json.MarshalIndent(applied, "", "  ")
	if err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to record: %s", err))
		return
	}
	if err := writeAtomic(currentPlanRevisionPath, string(data)); err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to record: %s", err))
		return
	}
	historyPath := filepath.Join(planRevisionsDir, fmt.Sprintf("plan-v%d.json", applied.Version))
	if err := writeAtomic(historyPath, string(data)); err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to record: %s", err))
	}
}

// HasPlanFingerprint reports whether the given fingerprint already appears in the revision
// history — dedupe guard against reapplying the identical plan in the same run.
func HasPlanFingerprint(fingerprint string) bool {
	if !dirExists(planRevisionsDir) {
		return false
	}
	entries, err := os.ReadDir(planRevisionsDir)
	if err != nil {
		return false
	}
	for _, entry := range entries {
		name := entry.Name()
		if entry.IsDir() || !strings.HasPrefix(name, "plan-v") || !strings.HasSuffix(name, ".json") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(planRevisionsDir, name))
		if err != nil {
			continue // corrupt history never approves a proposal
		}
		var applied AppliedPlanRevision
		if err := json.Unmarshal(data, &applied); err != nil {
			continue
		}
		if applied.PlanFingerprint == fingerprint {
			return true
		}
	}
	return false
}

// PlanRevisionFingerprint canonicalizes features (sorted by id, Passes forced false so a
// fingerprint doesn't change just because features later pass) and returns the lowercase hex
// SHA-256 of their JSON encoding.
func PlanRevisionFingerprint(features []Feature) string {
	canonical := make([]Feature, len(features))
	copy(canonical, features)
	sort.Slice(canonical, func(i, j int) bool { return canonical[i].Id < canonical[j].Id })
	for i := range canonical {
		canonical[i].Passes = false
	}

	data, err := json.Marshal(featureList{Items: canonical})
	if err != nil {
		LogError(fmt.Sprintf("[PlanRevisionStore] failed to fingerprint: %s", err))
		return ""
	}
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

// ResetPlanRevisions clears the proposal, current pointer and history — paired with
// ResetFeatures/ResetPlanObservations.
func ResetPlanRevisions() {
	if fileExists(ReplanProposalPath) {
		if err := os.Remove(ReplanProposalPath); err != nil {
			LogError(fmt.Sprintf("[PlanRevisionStore] failed to reset: %s", err))
		}
	}
	if fileExists(currentPlanRevisionPath) {
		if err := os.Remove(currentPlanRevisionPath); err != nil {
			LogError(fmt.Sprintf("[PlanRevisionStore] failed to reset: %s", err))
		}
	}
	if dirExists(planRevisionsDir) {
		if err := os.RemoveAll(planRevisionsDir); err != nil {
			LogError(fmt.Sprintf("[PlanRevisionStore] failed to reset: %s", err))
		}
	}
}
